backend/cmmgen.ml

```diff
@@ -195,7 +204,9 @@ let rec expr_size env = function
       expr_size env exp'
   | Uoffset (exp, offset) ->
       (match expr_size env exp with
-      | RHS_block blocksize -> RHS_infix { blocksize; offset }
+      | RHS_block (mode, blocksize) ->
+         assert_not_local mode;
```

Comment:
Since we allow mutually recursive functions to be local, I'd expect this assertion to fail from time to time.
Typically, for the following code:
```ocaml
let top cond zero =
  let rec f = local_ fun x -> if cond then zero else g x
  and g = local_ fun y -> f y in
  let r = if cond then f zero else g zero in
  r
```
I'd expect this code to fail this assertion at least with Closure and Flambda1.

Reply:
This particular example doesn't fail, because the uses of `Uoffset` aren't themselves used in a recursive definition. I've tried constructing an example that does fail, and I get errors about locals escaping their region. Given that the warning doesn't explain what region is being escaped, I don't know if my example is incorrect for some reason, or if it's a bug somewhere else.
Here is the updated version:
```ocaml
let top cond zero =
  let rec wrap =
    let rec f = local_ fun x -> if cond then zero else g x
    and g = local_ fun y -> f y in
    g
  in
  let r = if cond then wrap zero else zero in
  r
```

Reply:
It doesn't fail, because `expr_size` is not run on `let rec` bindings that contain only functions, as those generate a Clambda `Uclosure` (with several ufunctions), rather than a Clamda `Uletrec`. We only hit these cases when there is a `let rec` containing a mixture of mutually recursive closures and non-closure blocks, which cannot be locally allocated.

Reply:
In that case, what's the point of allowing a local mode in the code for `Uclosure` above ? As you said, it's only relevant if the local closures end up in an outer `Uletrec`, which should be forbidden ?

Reply:
@stedolan ping

Reply:
(I think this is resolved, see the last commit / comment)